File: android_mobile/src/main/java/com/alexstyl/specialdates/addevent/OperationsFactory.java
package com.alexstyl.specialdates.addevent;

import android.content.ContentProviderOperation;
import android.provider.ContactsContract;
import android.provider.ContactsContract.CommonDataKinds;
import android.provider.ContactsContract.CommonDataKinds.Photo;
import android.provider.ContactsContract.CommonDataKinds.StructuredName;
import android.provider.ContactsContract.Data;

import com.alexstyl.specialdates.contact.Contact;
import com.alexstyl.specialdates.date.ContactEvent;
import com.alexstyl.specialdates.date.Date;
import com.alexstyl.specialdates.events.Event;
import com.alexstyl.specialdates.events.peopleevents.EventType;
import com.alexstyl.specialdates.images.DecodedImage;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

final class OperationsFactory {

    private static final int NO_RAW_CONTACT_ID = 0;

    private final int rawContactID;

    static OperationsFactory forNewContact() {
        return new OperationsFactory(NO_RAW_CONTACT_ID);
    }

    OperationsFactory(int rawContactID) {
        this.rawContactID = rawContactID;
    }

    ContentProviderOperation newInsertFor(EventType eventType, Date date) {
        ContentProviderOperation.Builder builder = ContentProviderOperation
                .newInsert(Data.CONTENT_URI)
                .withValue(Data.MIMETYPE, CommonDataKinds.Event.CONTENT_ITEM_TYPE)
                .withValue(CommonDataKinds.Event.TYPE, eventType.getAndroidType())
                .withValue(CommonDataKinds.Event.START_DATE, date.toShortDate());
        addRawContactID(builder);
        return builder.build();
    }

    private void addRawContactID(ContentProviderOperation.Builder builder) {
        if (rawContactID == NO_RAW_CONTACT_ID) {
            builder.withValueBackReference(Data.RAW_CONTACT_ID, rawContactID);
        } else {
            builder.withValue(Data.RAW_CONTACT_ID, rawContactID);

        }
    }

    ArrayList<ContentProviderOperation> deleteEvents(List<ContactEvent> contactEvents) {
        ArrayList<ContentProviderOperation> ops = new ArrayList<>();
        for (ContactEvent contactEvent : contactEvents) {
            long eventId = contactEvent.getDeviceEventId().get();
            ops.add(
                    ContentProviderOperation
                            .newDelete(Data.CONTENT_URI)
                            .withSelection(CommonDataKinds.Event._ID + "= " + eventId, null)
                            .build());
        }
        return ops;
    }

    ArrayList<ContentProviderOperation> createContactIn(AccountData account, String contactName) {
        ArrayList<ContentProviderOperation> ops = new ArrayList<>(2);
        ops.add(
                ContentProviderOperation.newInsert(ContactsContract.RawContacts.CONTENT_URI)
                        .withValue(ContactsContract.RawContacts.ACCOUNT_NAME, account.getAccountName())
                        .withValue(ContactsContract.RawContacts.ACCOUNT_TYPE, account.getAccountType())
                        .build());
        ops.add(
                ContentProviderOperation.newInsert(Data.CONTENT_URI)
                        .withValueBackReference(Data.RAW_CONTACT_ID, rawContactID)
                        .withValue(Data.MIMETYPE, StructuredName.CONTENT_ITEM_TYPE)
                        .withValue(StructuredName.DISPLAY_NAME, contactName)
                        .build());

        return ops;
    }

    List<ContentProviderOperation> insertEvents(Collection<Event> events) {
        List<ContentProviderOperation> operations = new ArrayList<>(events.size());
        for (Event entry : events) {
            EventType eventType = entry.getEventType();
            Date date = entry.getDate();
            ContentProviderOperation operation = newInsertFor(eventType, date);
            operations.add(operation);
        }
        return operations;
    }

    ContentProviderOperation insertImageToContact(DecodedImage image) {
        ContentProviderOperation.Builder builder = ContentProviderOperation.newInsert(Data.CONTENT_URI)
                .withValue(Data.MIMETYPE, Photo.CONTENT_ITEM_TYPE)
                .withValue(Photo.PHOTO, image.getBytes());
        addRawContactID(builder);
        return builder.build();
    }

    ContentProviderOperation updateImageContact(DecodedImage decodedImage) {
        return ContentProviderOperation
                .newUpdate(ContactsContract.Data.CONTENT_URI)
                .withSelection(
                        Data.RAW_CONTACT_ID + "= ?" + " AND " + ContactsContract.Data.MIMETYPE + "=?",
                        new String[]{String.valueOf(rawContactID),
                                CommonDataKinds.Photo.CONTENT_ITEM_TYPE}
                )
                .withValue(CommonDataKinds.Photo.PHOTO, decodedImage.getBytes())
                .build();
    }

    ContentProviderOperation deleteImageFor(Contact contact) {
        return ContentProviderOperation.newDelete(Data.CONTENT_URI)
                .withValueBackReference(Data.RAW_CONTACT_ID, rawContactID)
                .withValue(Data.MIMETYPE, Photo.CONTENT_ITEM_TYPE)
                .withSelection(Photo.CONTACT_ID + "=?", new String[]{String.valueOf(contact.getContactID())})
                .build();
    }
}

File: android_mobile/src/main/java/com/alexstyl/specialdates/addevent/OperationsFactory.kt
package com.alexstyl.specialdates.addevent

import android.content.ContentProviderOperation
import android.provider.ContactsContract
import android.provider.ContactsContract.CommonDataKinds.Event
import android.provider.ContactsContract.CommonDataKinds.Photo
import android.provider.ContactsContract.CommonDataKinds.StructuredName
import android.provider.ContactsContract.Data
import com.alexstyl.specialdates.addevent.operations.ContactOperation
import com.alexstyl.specialdates.addevent.operations.InsertContact
import com.alexstyl.specialdates.addevent.operations.InsertEvent
import com.alexstyl.specialdates.addevent.operations.InsertImage
import com.alexstyl.specialdates.addevent.operations.UpdateContact
import com.alexstyl.specialdates.contact.Contact
import com.alexstyl.specialdates.contact.ImageURL
import com.alexstyl.specialdates.date.ContactEvent
import com.alexstyl.specialdates.date.Date
import com.alexstyl.specialdates.date.TimePeriod
import com.alexstyl.specialdates.events.database.EventTypeId
import com.alexstyl.specialdates.events.peopleevents.EventType
import com.alexstyl.specialdates.events.peopleevents.PeopleEventsProvider
import com.alexstyl.specialdates.events.peopleevents.ShortDateLabelCreator
import com.alexstyl.specialdates.events.peopleevents.StandardEventType
import com.alexstyl.specialdates.images.ImageDecoder

class OperationsFactory(private val rawContactID: Int,
                        private val displayStringCreator: ShortDateLabelCreator,
                        private val peopleEventsProvider: PeopleEventsProvider,
                        private val accountsProvider: WriteableAccountsProvider,
                        private val imageDecoder: ImageDecoder) {

    fun createOperationsFor(contactOperation: ContactOperation): List<ContentProviderOperation> {
        when (contactOperation) {
            is InsertContact -> return createContactIn(accountToStoreContact, contactOperation.contactName)
            is UpdateContact -> return updateExistingContact(contactOperation.contact)
            is InsertEvent -> return newInsertFor(contactOperation.eventType, contactOperation.date)
            is InsertImage -> return insertImageFor(contactOperation.imageUri)
        }
        throw IllegalArgumentException("Unable to create operation for $contactOperation")
    }

    private fun insertImageFor(imageUri: ImageURL): List<ContentProviderOperation> {
        val decodeFrom = imageDecoder.decodeFrom(imageUri)
        if (decodeFrom != null) {
            val builder = ContentProviderOperation.newInsert(Data.CONTENT_URI)
                    .withValue(Data.MIMETYPE, Photo.CONTENT_ITEM_TYPE)
                    .withValue(Photo.PHOTO, decodeFrom.bytes)
            addRawContactID(builder)
            return listOf(builder.build())
        } else {
            return emptyList()
        }
    }


    private fun newInsertFor(eventType: EventType, date: Date): List<ContentProviderOperation> {
        val builder = ContentProviderOperation
                .newInsert(Data.CONTENT_URI)
                .withValue(Data.MIMETYPE, Event.CONTENT_ITEM_TYPE)
                .withValue(Event.TYPE, androidIdOf(eventType))
                .withValue(Event.START_DATE, displayStringCreator.createLabelWithYearPreferredFor(date))
        addRawContactID(builder)
        return listOf(builder.build())
    }

    private fun androidIdOf(eventType: EventType): Int = when (eventType.id) {
        EventTypeId.TYPE_BIRTHDAY -> Event.TYPE_BIRTHDAY
        EventTypeId.TYPE_ANNIVERSARY -> Event.TYPE_ANNIVERSARY
        EventTypeId.TYPE_CUSTOM -> Event.TYPE_CUSTOM
        EventTypeId.TYPE_OTHER -> Event.TYPE_OTHER
        else -> {
            throw IllegalStateException("There is no Android type of $eventType")
        }
    }

    private fun addRawContactID(builder: ContentProviderOperation.Builder) {
        if (rawContactID == NO_RAW_CONTACT_ID) {
            builder.withValueBackReference(Data.RAW_CONTACT_ID, rawContactID)
        } else {
            builder.withValue(Data.RAW_CONTACT_ID, rawContactID)

        }
    }

    private fun deleteEvents(contactEvents: List<ContactEvent>): ArrayList<ContentProviderOperation> {
        val ops = ArrayList<ContentProviderOperation>()
        for (contactEvent in contactEvents) {
            val eventId = contactEvent.deviceEventId
            ops.add(
                    ContentProviderOperation
                            .newDelete(Data.CONTENT_URI)
                            .withSelection(Event._ID + "= " + eventId, null)
                            .build())
        }
        return ops
    }

    private fun createContactIn(account: AccountData, contactName: String): ArrayList<ContentProviderOperation> {
        val ops = ArrayList<ContentProviderOperation>(2)
        ops.add(
                ContentProviderOperation.newInsert(ContactsContract.RawContacts.CONTENT_URI)
                        .withValue(ContactsContract.RawContacts.ACCOUNT_NAME, account.accountName)
                        .withValue(ContactsContract.RawContacts.ACCOUNT_TYPE, account.accountType)
                        .build())
        ops.add(
                ContentProviderOperation.newInsert(Data.CONTENT_URI)
                        .withValueBackReference(Data.RAW_CONTACT_ID, rawContactID)
                        .withValue(Data.MIMETYPE, StructuredName.CONTENT_ITEM_TYPE)
                        .withValue(StructuredName.DISPLAY_NAME, contactName)
                        .build())

        return ops
    }


    private val accountToStoreContact: AccountData
        get() {
            val availableAccounts = accountsProvider.availableAccounts
            return if (availableAccounts.size == 0) {
                AccountData.NO_ACCOUNT
            } else {
                availableAccounts[0]
            }
        }

    private fun updateExistingContact(contact: Contact): List<ContentProviderOperation> {
        val contactEvents = getAllDeviceEventsFor(contact)
        return deleteEvents(contactEvents)
    }

    private fun getAllDeviceEventsFor(contact: Contact): List<ContactEvent> {
        val contactEvents = ArrayList<ContactEvent>()
        val contactEventsOnDate = peopleEventsProvider.fetchEventsBetween(TimePeriod.aYearFromNow())
        for (contactEvent in contactEventsOnDate) {
            val (contactID) = contactEvent.contact
            if (contactID == contact.contactID && contactEvent.type !== StandardEventType.NAMEDAY) {
                contactEvents.add(contactEvent)
            }
        }
        return contactEvents
    }

    companion object {

        private const val NO_RAW_CONTACT_ID = 0

        fun forNewContact(displayStringCreator: ShortDateLabelCreator,
                          peopleEventsProvider: PeopleEventsProvider,
                          accountsProvider: WriteableAccountsProvider,
                          imageDecoder: ImageDecoder): OperationsFactory {
            return OperationsFactory(NO_RAW_CONTACT_ID, displayStringCreator, peopleEventsProvider, accountsProvider,
                    imageDecoder)
        }
    }
}
